proxystore/store/dim/margo.py

```diff
@@ -0,0 +1,411 @@
+"""MargoStore implementation."""
+from __future__ import annotations
+
+import logging
+from multiprocessing import Process
+from time import sleep
+from typing import Any
+from typing import NamedTuple
+
+try:
+    import pymargo
+    import pymargo.bulk as bulk
+    from pymargo.core import Engine
+    from pymargo.core import Handle
+    from pymargo.core import RemoteFunction
+    from pymargo.bulk import Bulk
+
+    pymargo_import_error = None
+except ImportError as e:  # pragma: no cover
+    pymargo_import_error = e
+
+
+import proxystore.utils as utils
+from proxystore.store.base import Store
+from proxystore.store.dim.utils import get_ip_address
+from proxystore.store.dim.utils import Status
+
+ENCODING = 'UTF-8'
+
+server_process: Process | None = None
+logger = logging.getLogger(__name__)
+
+
+class MargoStoreKey(NamedTuple):
+    """Key to objects in a MargoStore."""
+
+    margo_key: str
+    obj_size: int
+    peer: str
+
+
+class MargoStore(Store[MargoStoreKey]):
+    """MargoStore implementation for intrasite communication."""
+
+    host: str
+    addr: str
+    protocol: str
+    engine: Engine
+    _rpc: dict[str, RemoteFunction]
+
+    # TODO : make host optional and try to get infiniband path automatically
+    def __init__(
+        self,
+        name: str,
+        *,
+        interface: str,
+        port: int,
+        protocol: str = 'verbs',
```

Comment:
Sure. I don't particularly like doing that because if mercury supports more protocols, we'll need to update our enum. There's nothing at this time I can borrow from py-mochi-margo, so I'll create it based on this figure https://mercury-hpc.github.io/user/na/ , but we might need to revert to strings eventually.